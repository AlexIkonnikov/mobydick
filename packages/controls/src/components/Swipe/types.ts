export interface ISwipe {
  active: boolean;
  disabled: boolean;
  onPress: (event: boolean) => void;
}
