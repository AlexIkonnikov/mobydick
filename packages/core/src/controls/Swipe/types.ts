export interface ISwipe {
  active: boolean;
  disabled: boolean;
  onPress: (isActive: boolean) => void;
}
