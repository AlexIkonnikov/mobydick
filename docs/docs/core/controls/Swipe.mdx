import ReactPlayer from 'react-player';
import Swipe from './assets/Swipe.mov';

<ReactPlayer playing controls url={Swipe} />

Reusable Swipe component

## Example

```tsx
import {useState} from 'react';

import {Swipe} from '@lad-tech/mobydick-core';

export const SwipeWidget = () => {
  const [active, setActive] = useState(false);
  return (
    <Swipe onPress={setActive} active={active} disabled={false} />
  );
};

```

## `Props`

### <div class="label required basic">Required</div>**`active`**

| TYPE    |
|:--------|
| boolean |

Manage on/off swipe

### <div class="label required basic">Required</div>**`disabled`**

| TYPE    |
|:--------|
| boolean |

Disable element

### <div class="label required basic">Required</div>**`onPress`**

| TYPE                        |
|:----------------------------|
| (isActive: boolean) => void |

Function for control active
