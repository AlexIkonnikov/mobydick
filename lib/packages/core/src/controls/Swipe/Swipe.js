"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const react_1 = require("react");
const react_native_1 = require("react-native");
const useStyles_1 = __importDefault(require("../../styles/hooks/useStyles"));
const other_1 = require("../../other");
const px_1 = __importDefault(require("../../styles/utils/px"));
const stylesCreate_1 = __importDefault(require("./stylesCreate"));
const leftPos = -20;
const rightPos = 20;
const Swipe = ({ active, disabled, onPress }) => {
    const [styles, theme] = (0, useStyles_1.default)(stylesCreate_1.default, disabled);
    const defaultState = active ? rightPos : leftPos;
    const pan = (0, react_1.useRef)(new react_native_1.Animated.Value(defaultState)).current;
    const distance = (0, react_1.useRef)(defaultState);
    (0, react_1.useEffect)(() => {
        if (active) {
            return swipeRight();
        }
        else {
            return swipeLeft();
        }
    }, [active, disabled]);
    const swipeRight = () => {
        distance.current = rightPos;
        react_native_1.Animated.timing(pan, {
            toValue: +rightPos,
            duration: 300,
            useNativeDriver: false,
            easing: react_native_1.Easing.ease,
        }).start();
    };
    const swipeLeft = () => {
        distance.current = leftPos;
        react_native_1.Animated.timing(pan, {
            toValue: +leftPos,
            duration: 300,
            useNativeDriver: false,
            easing: react_native_1.Easing.ease,
        }).start();
    };
    const handlePanResponderRelease = (_, gestureState) => {
        pan.flattenOffset();
        if (gestureState.dx === 0) {
            onPress(distance.current > 0 ? false : true);
            return distance.current > 0 ? swipeLeft() : swipeRight();
        }
        if (gestureState.dx > 0) {
            distance.current !== rightPos && onPress(true);
            return swipeRight();
        }
        else {
            distance.current !== leftPos && onPress(false);
            return swipeLeft();
        }
    };
    const panResponder = (0, react_1.useRef)(react_native_1.PanResponder.create({
        onStartShouldSetPanResponder: () => !disabled,
        onMoveShouldSetPanResponder: () => !disabled,
        onPanResponderGrant: () => {
            pan.setOffset(distance.current);
        },
        onPanResponderMove: react_native_1.Animated.event([null, { dx: pan }], {
            useNativeDriver: false,
        }),
        onPanResponderRelease: handlePanResponderRelease,
    })).current;
    const backgroundColor = pan.interpolate({
        inputRange: [leftPos, rightPos],
        outputRange: [theme.colors.ElementMuted, theme.colors.ElementBase],
        extrapolate: 'clamp',
    });
    const translateX = pan.interpolate({
        inputRange: [leftPos, rightPos],
        outputRange: [0, (0, px_1.default)(20)],
        extrapolate: 'clamp',
    });
    return (<react_native_1.Animated.View style={[styles.container, { backgroundColor }]} needsOffscreenAlphaCompositing={true} accessibilityLabel={other_1.LABELS.swipe} {...panResponder.panHandlers}>
      <react_native_1.Animated.View style={[styles.switcher, { transform: [{ translateX }] }]}/>
    </react_native_1.Animated.View>);
};
exports.default = Swipe;
