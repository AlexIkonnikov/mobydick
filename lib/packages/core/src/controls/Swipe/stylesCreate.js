"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const styles_1 = require("../../styles");
const px_1 = __importDefault(require("../../styles/utils/px"));
const stylesCreate = (0, styles_1.createStyles)(({ spaces, colors }, disabled) => ({
    container: {
        width: (0, px_1.default)(50),
        height: (0, px_1.default)(30),
        borderRadius: spaces.Space20,
        padding: spaces.Space2,
        opacity: disabled ? 0.4 : 1,
    },
    switcher: {
        flex: 1,
        aspectRatio: 1,
        backgroundColor: colors.ElementWhite,
        borderRadius: (0, px_1.default)(25),
    },
}));
exports.default = stylesCreate;
