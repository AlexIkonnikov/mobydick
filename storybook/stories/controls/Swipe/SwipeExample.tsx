import React, {useCallback, useState} from 'react';

import {Button, Swipe, View} from '@lad-tech/mobydick-core';

const SwipeExample = () => {
  const [isActive, setActive] = useState(false);
  const [isActiveTwo, setActiveTwo] = useState(false);

  const onPressCommon = useCallback(() => {
    setActive(isActive => !isActive);
    setActiveTwo(isActiveTwo => !isActiveTwo);
  }, []);

  const onPressOne = useCallback((isActive: boolean) => {
    setActive(isActive);
  }, []);

  const onPressTwo = useCallback(() => {
    setActiveTwo(isActiveTwo => !isActiveTwo);
  }, []);

  return (
    <>
      <Swipe onPress={onPressOne} active={isActive} disabled={false} />
      <View style={{padding: 20}}>
        <Swipe onPress={onPressTwo} active={isActiveTwo} disabled={true} />
      </View>
      <Button onPress={onPressCommon} text={'Изменить состояние'} />
    </>
  );
};

export default SwipeExample;
