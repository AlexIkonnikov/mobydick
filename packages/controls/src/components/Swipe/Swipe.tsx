import React, {FC, useEffect, useRef, useState} from 'react';
import {rem, useStyles} from '@npm/mobydick-styles';
import {Animated, Easing, PanResponder} from 'react-native';

import stylesCreate from '../Toggle/stylesCreate';

import {ISwipe} from './types';

const leftPos = -20;
const rightPos = 20;

const Swipe: FC<ISwipe> = ({active, disabled, onPress}) => {
  const [styles, theme] = useStyles(stylesCreate, disabled);
  const defaultState = active ? rightPos : leftPos;

  const pan = useRef(new Animated.Value(defaultState)).current;
  const distance = useRef<number>(defaultState);

  const [isActive, setActive] = useState(active);

  useEffect(() => {
    onPress(isActive);
  }, [isActive]);

  const swipeRight = () => {
    distance.current = rightPos;
    setActive(true);
    Animated.timing(pan, {
      toValue: +rightPos,
      duration: 300,
      useNativeDriver: false,
      easing: Easing.ease,
    }).start();
  };

  const swipeLeft = () => {
    distance.current = leftPos;
    setActive(false);
    Animated.timing(pan, {
      toValue: +leftPos,
      duration: 300,
      useNativeDriver: false,
      easing: Easing.ease,
    }).start();
  };

  const panResponder = useRef(
    PanResponder.create({
      onStartShouldSetPanResponder: () => !disabled,
      onMoveShouldSetPanResponder: () => !disabled,

      onPanResponderGrant: () => {
        pan.setOffset(distance.current);
      },

      onPanResponderMove: Animated.event([null, {dx: pan}], {
        useNativeDriver: false,
      }),

      onPanResponderRelease: (_event, gestureState) => {
        pan.flattenOffset();

        if (gestureState.dx === 0) {
          if (distance.current > 0) {
            return swipeLeft();
          } else {
            return swipeRight();
          }
        }

        distance.current = gestureState.dx;
        if (distance.current > 0) {
          return swipeRight();
        } else {
          return swipeLeft();
        }
      },
    }),
  ).current;

  const backgroundColor = pan.interpolate({
    inputRange: [leftPos, rightPos],
    outputRange: [theme.colors.ElementMuted, theme.colors.ElementBase],
    extrapolate: 'clamp',
  });

  const translateX = pan.interpolate({
    inputRange: [leftPos, rightPos],
    outputRange: [0, rem(20)],
    extrapolate: 'clamp',
  });

  return (
    <Animated.View
      style={[styles.container, {backgroundColor}]}
      needsOffscreenAlphaCompositing={true}
      {...panResponder.panHandlers}>
      <Animated.View style={[styles.switcher, {transform: [{translateX}]}]} />
    </Animated.View>
  );
};

export default Swipe;
