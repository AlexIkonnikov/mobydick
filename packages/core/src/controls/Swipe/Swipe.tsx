import React, {FC, useEffect, useRef} from 'react';
import {
  Animated,
  Easing,
  GestureResponderEvent,
  PanResponder,
  PanResponderGestureState,
} from 'react-native';

import useStyles from '../../styles/theme/hooks/useStyles';
import rem from '../../styles/spaces/rem';
import {LABELS} from '../../other';

import stylesCreate from './stylesCreate';
import {ISwipe} from './types';

const leftPos = -20;
const rightPos = 20;

const Swipe: FC<ISwipe> = ({active, disabled, onPress}) => {
  const [styles, theme] = useStyles(stylesCreate, disabled);
  const defaultState = active ? rightPos : leftPos;

  const pan = useRef(new Animated.Value(defaultState)).current;
  const distance = useRef<number>(defaultState);

  useEffect(() => {
    if (disabled) {
      return;
    }
    if (active) {
      return swipeRight();
    } else {
      return swipeLeft();
    }
  }, [active, disabled]);

  const swipeRight = () => {
    distance.current = rightPos;
    onPress(true);
    Animated.timing(pan, {
      toValue: +rightPos,
      duration: 300,
      useNativeDriver: false,
      easing: Easing.ease,
    }).start();
  };

  const swipeLeft = () => {
    distance.current = leftPos;
    onPress(false);
    Animated.timing(pan, {
      toValue: +leftPos,
      duration: 300,
      useNativeDriver: false,
      easing: Easing.ease,
    }).start();
  };

  const handlePanResponderRelease = (
    _: GestureResponderEvent,
    gestureState: PanResponderGestureState,
  ) => {
    pan.flattenOffset();

    if (gestureState.dx === 0) {
      if (distance.current > 0) {
        return swipeLeft();
      } else {
        return swipeRight();
      }
    }

    distance.current = gestureState.dx;
    if (distance.current > 0) {
      return swipeRight();
    } else {
      return swipeLeft();
    }
  };

  const panResponder = useRef(
    PanResponder.create({
      onStartShouldSetPanResponder: () => !disabled,
      onMoveShouldSetPanResponder: () => !disabled,

      onPanResponderGrant: () => {
        pan.setOffset(distance.current);
      },

      onPanResponderMove: Animated.event([null, {dx: pan}], {
        useNativeDriver: false,
      }),

      onPanResponderRelease: handlePanResponderRelease,
    }),
  ).current;

  const backgroundColor = pan.interpolate({
    inputRange: [leftPos, rightPos],
    outputRange: [theme.colors.ElementMuted, theme.colors.ElementBase],
    extrapolate: 'clamp',
  });

  const translateX = pan.interpolate({
    inputRange: [leftPos, rightPos],
    outputRange: [0, rem(20)],
    extrapolate: 'clamp',
  });

  return (
    <Animated.View
      style={[styles.container, {backgroundColor}]}
      needsOffscreenAlphaCompositing={true}
      accessibilityLabel={LABELS.swipe}
      {...panResponder.panHandlers}>
      <Animated.View style={[styles.switcher, {transform: [{translateX}]}]} />
    </Animated.View>
  );
};

export default Swipe;
