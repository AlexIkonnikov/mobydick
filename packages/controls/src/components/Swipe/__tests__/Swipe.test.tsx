import {act, render} from '@testing-library/react-native';
import React from 'react';

import Swipe from '../Swipe';

const onPress = () => null;
// const gestureStateMOck = {
//   dx: 0,
//   dy: 0,
//   moveX: 0,
//   moveY: 0,
//   numberActiveTouches: 0,
//   stateID: 0.36193975836547276,
//   vx: 0,
//   vy: 0,
//   x0: 0,
//   y0: 0,
// };
//
// const gestureStateMOckDx = {
//   dx: 20,
//   dy: 0,
//   moveX: 0,
//   moveY: 0,
//   numberActiveTouches: 0,
//   stateID: 0.36193975836547276,
//   vx: 0,
//   vy: 0,
//   x0: 0,
//   y0: 0,
// };
const eventMock = {};
describe('Swipe', () => {
  it('should renders correctly by default -', function () {
    const {getByLabelText, toJSON} = render(
      <Swipe active={false} disabled={false} onPress={onPress} />,
    );
    const panHandler = getByLabelText('swipe');
    act(() => panHandler.props.onPanResponderRelease({}, {dx: -10}));
    expect(toJSON()).toMatchSnapshot();
  });
  it('should renders correctly by default + ', function () {
    const {getByLabelText, toJSON} = render(
      <Swipe active={false} disabled={false} onPress={onPress} />,
    );
    const panHandler = getByLabelText('swipe');
    act(() => panHandler.props.onPanResponderRelease({}, {dx: 20}));
    expect(toJSON()).toMatchSnapshot();
  });
  it('should renders correctly active', function () {
    const {toJSON} = render(
      <Swipe active={false} disabled={false} onPress={onPress} />,
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('should renders correctly disabled + ', function () {
    const {getByLabelText, toJSON} = render(
      <Swipe active={true} disabled onPress={onPress} />,
    );
    const panHandler = getByLabelText('swipe');

    act(() => {
      panHandler.props.onStartShouldSetPanResponder(() => true);
      panHandler.props.onMoveShouldSetPanResponder(() => true);
      panHandler.props.onPanResponderGrant();
      panHandler.props.onPanResponderMove(eventMock, {dx: 0});
      panHandler.props.onPanResponderRelease(eventMock, {dx: 20});
    });
    expect(toJSON()).toMatchSnapshot();
  });
  it('should renders correctly disabled - ', function () {
    const {getByLabelText, toJSON} = render(
      <Swipe active={true} disabled onPress={onPress} />,
    );
    const panHandler = getByLabelText('swipe');

    act(() => {
      panHandler.props.onStartShouldSetPanResponder(() => true);
      panHandler.props.onMoveShouldSetPanResponder(() => true);
      panHandler.props.onPanResponderGrant();
      panHandler.props.onPanResponderMove(eventMock, {dx: 0});
      panHandler.props.onPanResponderRelease(eventMock, {dx: -20});
    });
    expect(toJSON()).toMatchSnapshot();
  });
  it('should renders correctly active and disabled', function () {
    const {toJSON} = render(<Swipe active disabled onPress={onPress} />);
    expect(toJSON()).toMatchSnapshot();
  });
});
