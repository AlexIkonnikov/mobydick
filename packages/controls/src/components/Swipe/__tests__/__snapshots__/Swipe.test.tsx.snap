// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Swipe should renders correctly active 1`] = `
<View
  accessibilityLabel="swipe"
  collapsable={false}
  needsOffscreenAlphaCompositing={true}
  onMoveShouldSetPanResponder={[Function]}
  onPanResponderGrant={[Function]}
  onPanResponderMove={[Function]}
  onPanResponderRelease={[Function]}
  onStartShouldSetPanResponder={[Function]}
  style={
    Object {
      "backgroundColor": "rgba(155, 161, 176, 1)",
      "borderRadius": 30,
      "height": 45,
      "opacity": 1,
      "padding": 3,
      "width": 75,
    }
  }
>
  <View
    collapsable={false}
    style={
      Object {
        "aspectRatio": 1,
        "backgroundColor": "#FFF",
        "borderRadius": 37.5,
        "flex": 1,
        "transform": Array [
          Object {
            "translateX": 0,
          },
        ],
      }
    }
  />
</View>
`;

exports[`Swipe should renders correctly active and disabled 1`] = `
<View
  accessibilityLabel="swipe"
  collapsable={false}
  needsOffscreenAlphaCompositing={true}
  onMoveShouldSetPanResponder={[Function]}
  onPanResponderGrant={[Function]}
  onPanResponderMove={[Function]}
  onPanResponderRelease={[Function]}
  onStartShouldSetPanResponder={[Function]}
  style={
    Object {
      "backgroundColor": "rgba(43, 120, 238, 1)",
      "borderRadius": 30,
      "height": 45,
      "opacity": 0.4,
      "padding": 3,
      "width": 75,
    }
  }
>
  <View
    collapsable={false}
    style={
      Object {
        "aspectRatio": 1,
        "backgroundColor": "#FFF",
        "borderRadius": 37.5,
        "flex": 1,
        "transform": Array [
          Object {
            "translateX": 30,
          },
        ],
      }
    }
  />
</View>
`;

exports[`Swipe should renders correctly by default +  1`] = `
<View
  accessibilityLabel="swipe"
  collapsable={false}
  needsOffscreenAlphaCompositing={true}
  onMoveShouldSetPanResponder={[Function]}
  onPanResponderGrant={[Function]}
  onPanResponderMove={[Function]}
  onPanResponderRelease={[Function]}
  onStartShouldSetPanResponder={[Function]}
  style={
    Object {
      "backgroundColor": "rgba(43, 120, 238, 1)",
      "borderRadius": 30,
      "height": 45,
      "opacity": 1,
      "padding": 3,
      "width": 75,
    }
  }
>
  <View
    collapsable={false}
    style={
      Object {
        "aspectRatio": 1,
        "backgroundColor": "#FFF",
        "borderRadius": 37.5,
        "flex": 1,
        "transform": Array [
          Object {
            "translateX": 30,
          },
        ],
      }
    }
  />
</View>
`;

exports[`Swipe should renders correctly by default - 1`] = `
<View
  accessibilityLabel="swipe"
  collapsable={false}
  needsOffscreenAlphaCompositing={true}
  onMoveShouldSetPanResponder={[Function]}
  onPanResponderGrant={[Function]}
  onPanResponderMove={[Function]}
  onPanResponderRelease={[Function]}
  onStartShouldSetPanResponder={[Function]}
  style={
    Object {
      "backgroundColor": "rgba(155, 161, 176, 1)",
      "borderRadius": 30,
      "height": 45,
      "opacity": 1,
      "padding": 3,
      "width": 75,
    }
  }
>
  <View
    collapsable={false}
    style={
      Object {
        "aspectRatio": 1,
        "backgroundColor": "#FFF",
        "borderRadius": 37.5,
        "flex": 1,
        "transform": Array [
          Object {
            "translateX": 0,
          },
        ],
      }
    }
  />
</View>
`;

exports[`Swipe should renders correctly disabled +  1`] = `
<View
  accessibilityLabel="swipe"
  collapsable={false}
  needsOffscreenAlphaCompositing={true}
  onMoveShouldSetPanResponder={[Function]}
  onPanResponderGrant={[Function]}
  onPanResponderMove={[Function]}
  onPanResponderRelease={[Function]}
  onStartShouldSetPanResponder={[Function]}
  style={
    Object {
      "backgroundColor": "rgba(43, 120, 238, 1)",
      "borderRadius": 30,
      "height": 45,
      "opacity": 0.4,
      "padding": 3,
      "width": 75,
    }
  }
>
  <View
    collapsable={false}
    style={
      Object {
        "aspectRatio": 1,
        "backgroundColor": "#FFF",
        "borderRadius": 37.5,
        "flex": 1,
        "transform": Array [
          Object {
            "translateX": 30,
          },
        ],
      }
    }
  />
</View>
`;

exports[`Swipe should renders correctly disabled -  1`] = `
<View
  accessibilityLabel="swipe"
  collapsable={false}
  needsOffscreenAlphaCompositing={true}
  onMoveShouldSetPanResponder={[Function]}
  onPanResponderGrant={[Function]}
  onPanResponderMove={[Function]}
  onPanResponderRelease={[Function]}
  onStartShouldSetPanResponder={[Function]}
  style={
    Object {
      "backgroundColor": "rgba(155, 161, 176, 1)",
      "borderRadius": 30,
      "height": 45,
      "opacity": 0.4,
      "padding": 3,
      "width": 75,
    }
  }
>
  <View
    collapsable={false}
    style={
      Object {
        "aspectRatio": 1,
        "backgroundColor": "#FFF",
        "borderRadius": 37.5,
        "flex": 1,
        "transform": Array [
          Object {
            "translateX": 0,
          },
        ],
      }
    }
  />
</View>
`;
